e2e storage tests: usable out-of-tree

pohly opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Not all CSI drivers can be tested in Kubernetes (they might not be open) or should not be (if it does not help develop and enhance Kubernetes). Therefore it is useful to make the tests available to out-of-tree drivers without pulling in Kubernetes E2E specific code and tests.

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
NONE
```

pohly committed:
e2e: update bazel BUILD files

Generated via hack/update-bazel.sh.

pohly committed:
e2e/storage: drivers determine the claim size

Different drivers support different volume sizes. Some have certain
minimum sizes, some maximum sizes. Instead of hard-coding some kind of
default into the testsuites, now each driver that supports dynamic
provisioning has to provide the size.

pohly committed:
e2e/storage: separate generic testsuites from specific drivers

This makes it possible to use the testsuites package out-of-tree
without pulling in unnecessary dependencies and code (in
test/e2e/storage/vsphere) that defines tests that are not wanted in a
custom test suite.

pohly commented:
/test pull-kubernetes-e2e-gce-100-performance

pohly committed:
e2e/storage: refactor test config

Exposing framework.VolumeTestConfig as part of the testsuite package
API was confusing because it was unclear which of the values in it
really have an effect. How it was set also was a bit awkward: a test
driver had a copy that had to be overwritten at test runtime and then
might have been updated and/or overwritten again by the driver.

Now testsuites has its own test config structure. It contains the
values that might have to be set dynamically at runtime. Instead of
overwriting a copy of that struct inside the test driver, the test
driver takes some common defaults (specifically, the framework pointer
and the prefix) when it gets initialized and then manages its own
copy. For example, the hostpath driver has to lock the pods to a
single node.

framework.VolumeTestConfig is still used internally and test drivers
can decide to run tests with a fully populated instance if needed (for
example, after setting up an NFS server).

pohly commented:
/hold

I am still testing this myself...

pohly commented:
/hold cancel

@msau42 @mkimuram please take a look. With these changes I can use the volumes testsuites out-of-tree. I ended up making some changes to enhance the experience.

Something that we need to check carefully now is that the same set of tests runs as before. The situation for the gce job (https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/70862/pull-kubernetes-e2e-gce/59623/) is that a single "CSI" test ran:
```
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] volumes should be mountable
```

Several tests were skipped, probably because of the `[Serial]` tag:
````
[sig-storage] CSI Volumes [Driver: com.google.csi.gcepd][Serial] [Testpattern: Dynamic PV (block volmode)] volumeMode[Feature:BlockVolume] should fail in binding dynamic provisioned PV to PVC
[sig-storage] CSI Volumes [Driver: com.google.csi.gcepd][Serial] [Testpattern: Dynamic PV (default fs)] provisioning should create and delete block persistent volumes [Feature:BlockVolume]
....
````

Does that look normal?

pohly commented:
/test pull-kubernetes-kubemark-e2e-gce-big

mkimuram commented:
@pohly 

Thank you for working on improvement for e2e testing framework.

I see many improvements on other than just making it "usable out-of-tree", like skipping tests efficiently. (They look really nice.) However, could you let me know which exact commits are necessary for this purpose? (I guess that only first one and second one would be needed, if so we would be better to make others to the separate PR.)

And could you share how you could make it use out-of-tree after the changes, so that anyone can confirm that it can be applied to other csi drivers by the changes?

> Several tests were skipped, probably because of the [Serial] tag:
> Does that look normal?

Yes. I think that tests tagged with [Serial], [Slow], or [Feature:*] are skipped in this job.
(There might be some more checks.)

pohly commented:
/assign @mkimuram

pohly commented:
I have split out some of the commits into separate PRs:
- fmt cleanup: PR #70989 (still included here to avoid conflicts, so if we merge this first, #70989 can be closed)
- test names: PR #70987 (can be merged independently)
- avoid defining unusable tests: PR #70992

pohly commented:
/test pull-kubernetes-integration

mkimuram commented:
@pohly 

Thank you for your explanation and separating PR. Now we can focus on making csi e2e-test out-of-tree in this PR.

Could you explain more about the relationship of the comment in ```out-of-tree e2e-test``` code and
```"e2e storage tests: usable out-of-tree"``` commit?
It doesn't still clear to me what makes difficult to make existing csi e2e-test code use out-of-tree and 
why moving definition of ```TestDriver``` class from ```k8s.io/kubernetes/test/e2e/storage/drivers``` to ```k8s.io/kubernetes/test/e2e/storage/testsuites/testdriver``` solves the issue.

Isn't it enough just importing necessary packages and just enabling necessary drivers in out-of-tree test code side?
(If you could share your working code, instead of making the main part of the codes blank and writing a comment, it would help me understanding your intention.)

- [comment in out-of-tree e2e-test code](https://github.com/pohly/csi-e2e/blob/master/test/e2e/e2e_test.go#L28)
```
	// test sources
	// The goal is to import the CSI storage tests. Right now
	// we can't do that because that pulls in also all of the other
	// storage tests and their dependencies, which we don't want.
	// _ "k8s.io/kubernetes/test/e2e/storage"
```

- ["e2e storage tests: usable out-of-tree" commit](https://github.com/kubernetes/kubernetes/pull/70862/commits/ff748a3ae49d3849013bc3b04b23ed0c1cfc4fc6)
```
e2e storage tests: usable out-of-tree

e2e/storage: separate generic testsuites from specific drivers
This makes it possible to use the testsuites package out-of-tree
without pulling in unnecessary dependencies and code (in
test/e2e/storage/vsphere) that defines tests that are not wanted in a
custom test suite.
```

pohly commented:
@mkimuram the "Right now we can't do that because that pulls in also all of the other storage tests and their dependencies, which we don't want." comment is old and refers to the `test/e2e/storage` package as it existed before the introduction of the `testsuites` package.

However, the underlying issue is the same: someone who wants to use the tests out-of-tree is not necessarily interested in running *all* of them in exactly the same way as the in-tree `test/e2e/storage` package does.

Consider a CSI driver developer: that person will want to set up an empty E2E test suite with minimal dependencies. The E2E framework refactoring made that possible (current state of csi-e2e demonstrates that).

But an empty test suite isn't very useful, so the next step is to import some tests. Your `testsuites` work is a very good first step towards that, but in it's current form it has the drawback that importing `test/e2e/storage/testsuites` imports `test/e2e/storage/drivers`, which pulls in additional, otherwise unnecessary packages. Worse, one of those dependencies (the vsphere package) adds tests to the test  suite into which the package gets linked, without a way to disable that.

Importing unnecessary packages causes overhead (as in "dep runs slower" and "vendor directory becomes larger") or can make vendoring so complicated that it fails entirely (figuring out exactly the right versions of different dependencies can be non-trivial).

Carefully separating code into layered packages as this pr (`test/e2e/storage/drivers` -> `test/e2e/storage/testsuites`) makes it possible to import the latter without the in-tree test drivers and their dependencies.

Tomorrow I can update the `csi-e2e` repo using two different Kubernetes branches: current master and with this PR. That should demonstrate the difference.

pohly commented:
@mkimuram I've updated https://github.com/pohly/csi-e2e. You can see the difference in dependencies by looking at the number of removed packages in the latest commit, which is the one using this PR instead of master.

This PR includes commits ("drivers determine the claim size " and "refactor test config") that enhance the test configuration. This isn't necessary for the dependency issue, but it is important to present an easy-to-use and understand API for external developers. IMHO it's not worthwhile to merge either of these separately because both should be merged before this can be recommended to external developers.

pohly commented:
/test pull-kubernetes-integration

mkimuram commented:
@pohly 

Thank you for sharing the code. It looks really good that we seem almost ready to use this framework out-fo-tree.

Is my understanding right that the reason that ```storage/vsphere``` is pulled is because ```storage/drivers/in_tree.go```  is [directly calling methods defined in ```storage/vsphere```](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1312)?

If so, isn't it better to fix it by defining wrapper in ```framework``` and call methods defined in ```storage/vsphere``` from it, as you did in https://github.com/kubernetes/kubernetes/commit/8b17db7e0c4431cd5fd9a5d9a3ab11b04e2f0a7e#diff-a7e923dd9f9d2690ade65713d2706137 ?

I'm more inclined to the idea of separating the definition of ```TestDriver``` and implementation of it than before. However, I still would like to keep drivers and testsuites separate, because once this separation breaks testsuites will be un-reusable. (I'm not sure if moving it to testsuites/testdriver really affects.) Also, it looks a bit strange that implementation of ```TestDriver``` in k/k still exists in ```storage/drivers```, but the definition of ```TestDriver``` are in ```storage/testsuites/testdrivers```, or a complete different place.

pohly commented:
Masaki Kimura <notifications@github.com> writes:
> Is my understanding right that the reason that ```storage/vsphere```
> is pulled is because ```storage/drivers/in_tree.go``` is [directly
> calling methods defined in
> ```storage/vsphere```](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1312)?

Correct.

> If so, isn't it better to fix it by defining wrapper in
> ```framework``` and call methods defined in ```storage/vsphere``` from
> it, as you did in
> https://github.com/kubernetes/kubernetes/commit/8b17db7e0c4431cd5fd9a5d9a3ab11b04e2f0a7e#diff-a7e923dd9f9d2690ade65713d2706137
> ?

I don't see why this should be better. It just makes the framework more
complicated. There is already an abstract API for calling
driver-specific functionality (`TestDriver`), there's no need to add
another one in the framework.

But perhaps I am missing something. What advantage does that have over
not importing the vsphere test driver (or any other of the in-tree test
drivers)?  Remember, I don't want to use those test drivers in my test
suite.

> I'm more inclined to the idea of separating the definition of
> ```TestDriver``` and implementation of it than before. However, I
> still would like to keep drivers and testsuites separate, because once
> this separation breaks testsuites will be un-reusable.

And this PR keeps them separate, doesn't it?
`tests/e2e/storages/testsuites` defines the interface (in a sub-package,
but it could also be in tests/e2e/storages/testsuites itself - this is
debatable with arguments both ways), and `tests/e2e/storage/drivers`
contains some implementations of it.

> (I'm not sure
> if moving it to testsuites/testdriver really affects.) Also, it looks
> a bit strange that implementation of ```TestDriver``` in k/k still
> exists in ```storage/drivers```, but the definition of
> ```TestDriver``` are in ```storage/testsuites/testdrivers```, or a
> complete different place.

We can also move `storage/drivers` to `storage/testsuites/drivers`, as
long as it is a completely separate package and `storage/testsuites`
doesn't depend on it. But IMHO it makes more sense to keep it out of
that directory, just as the definition of other test drivers will also be
elsewhere.

pohly commented:
/test pull-kubernetes-integration

msau42 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -48,6 +48,9 @@ type StorageClassTest struct {
 	PvCheck            func(volume *v1.PersistentVolume) error
 	NodeName           string
+	NodeSelector       map[string]string
+	SecondNodeName     string
```
Are there any tests that actually need a second node to be specified?

pohly replied:
The test that runs two pods runs the second pod on the second node. That way one can test moving a volume from one node to another.

It would be better to have more tests that involve various combinations of multiple pods and multiple nodes, like trying to use the same volume on the same node by two pods, or trying to use the same volume on two different nodes at the same time. But at least a single test is a start, and all of those additional tests would depend on a config option for the node(s) that they should use.

msau42 replied:
This nodeselector was mainly intended for hostpath volumes which can only be scheduled to a specific node. It's also used by multiple tests, not just the provisioning ones.  For plugins like PD, we don't want to set a nodeselector because we want to test the volume scheduling capabilities.

I think the test itself can dynamically construct a node affinity that excludes the node the first pod was scheduled to.  However, something like hostpath or local volumes won't work with this kind of test so it may need to define a plugin capability.

pohly replied:
For PD it's fine to leave all of these fields empty.

Same with the SecondNodeName: the hostpath driver has to leave it empty for correct execution.

msau42 replied:
I'm not sure if I want to make SecondNodeName a driver configurable parameter.

I'm thinking in the future, we will want a conformance test that explicitly moves the pod to a different node.  For such a conformance test, this second node should be controlled by the test case itself, not an optional argument set by a driver.

Plugins that don't support multi-node, like hostpath or local, will not be able to pass this test and we can add a driver capability to skip it.

pohly replied:
IMHO such a conformance test can be written without changing `TestConfig`. But when we need to make changes for such a conformance test, we can still make them later, okay?

msau42 replied:
I'm actually thinking we could reuse this same test and promote it as a conformance test.  I don't think we should need to write a separate test just for conformance that does basically the same function as a non-conformance test.  I also don't want to have to remove this field in the future and break people that are vendoring this test suite.

This second node setting, to me, seems more appropriate to drive from the test case itself, and not the driver.  The driver should be able to define its topology constraints, but shouldn't be forcing scheduling behavior.  Setting nodename in the Pod bypasses the scheduler and eliminates that whole path from e2e testing.  I agree that current tests are doing this for hostpath, but I want to refactor the old tests to move away from that model, and move towards a model where we handle topology more holistically.

This new field is adding new functionality to a test and is not part of the refactoring.  I think we should consider enhancing this test case as a separate PR.

msau42 reviewed (commented):
This generally lgtm, but I want to wait on merging this until all the required CSI GA work is merged.

mkimuram commented:
@pohly 

After I heard your explanation, I started to think that we could do like below. (Also see details.)
My intentions are:
- (1) Move all definition under ```lib/``` to make out-of-tree code only require this at minimum,
- (2) keep implementation of testsuites and drivers as they are,
- (3) Move common logic of test execution logic to ```lib.``` to make it available for out-of-tree test code,
- (4) Make in_tree_volumes.go and csi_volumes.go just call (3) with (2) to keep k/k test code and out-of-tree test code use the same logic. (out-of-tree may call (3) with just its own driver and testsuites implementation of (1), or may call (3) with some of (2), or combination of them.)

Also, We might be able to move ```drivers``` under ```testsuites``` as you suggested, in this case, too. And names in below details are just an example, so they could be different.

How do you think of this idea? (Is it too much?)

<details>

[Before]
```
storage/
        drivers/
               base.go
               in_tree.go
               csi.go
        testpatterns/
               testpattern.go
        testsuites/
               base.go
               volumes.go
               etc ...
        in_tree_volumes.go
        csi_volumes.go
```

[After]
```
storage/
        lib/
           drivers/base.go       (Moved from storage/drivers/base.go)
           testsuites/base.go    (Moved from storage/testsuites/base.go)
           xxx.go                (Common logic is moved from storage/in_tree_volumes.go and csi_volumes.go)
           testpattern.go        (Moved from storage/testpatterns/testpatterns.go and could be merged with xxx.go above)
        drivers/
               in_tree.go
               csi.go
        testsuites/
               volumes.go
               etc ...
        in_tree_volumes.go
        csi_volumes.go
```
</details>

mkimuram commented on test/e2e/storage/drivers/base.go:
```diff
@@ -74,6 +74,9 @@ type DynamicPVTestDriver interface {
 	GetDynamicProvisionStorageClass(fsType string) *storagev1.StorageClass
+
+	// GetClaimSize returns the size of the volume that is to be provisioned ("5Gi", "1Mi").
+	GetClaimSize() string
```
Maximum size that a driver can return depends on the driver.
On the other hand, minimum size that a test requires depends on the test.

Please see [volumeIO test](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/volume_io.go#L304), [driver definition](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/base.go#L84), and [how it was arranged](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/volume_io.go#L84). I prefer to do similar for it. (We could skip the test if a driver can't support the size that a test requires.)

If we are not going to add tests that the size matter in the future, it could be a over spec, though.
Also, if we would like to choose a short term solution, just wait for latest csi gce driver would be good, then we can change it back to 1Gi, which will fit to most drivers.

pohly replied:
So how about we let the driver define a minimum and maximum size for volumes that they can provision, and then tests can pick any size they like in that range when they provision volumes dynamically? Most tests don't care about the size at all and can just pick the minimum.

msau42 replied:
Some drivers also have a min size

pohly replied:
@msau42 yes, that's why I suggested defining a range ("define a minimum and maximum size for volumes").

pohly replied:
@mkimuram while looking at how min/max size information could be used by a test, I ran into a snag:
https://github.com/kubernetes/kubernetes/blob/e9aabf97e1d969ed80719cb01d8acb4f2be221b7/test/e2e/storage/testsuites/base.go#L138-L169

That function must dynamically provision a volume, but it has no information about the size expected by the test because conceptually, that code is independent from any particular test.

Does it really make sense to specify a minimum and maximum size? I'm more leaning towards just letting the driver suggest a default size which may be larger than the absolute minimum supported by the storage system.

Optionally, tests can check whether the volume is large enough and skip otherwise. Which tests depend on a certain minimum size of the volume, and can they determine what the size of volume they operate against is? However, I would prefer to keep this out of this PR. Instead I would update the GetClaimSize comment explaining that the size must be chosen so that it is large enough for all tests that are intended to pass.

verult replied:
@pohly with a single default size from the driver, how does the framework decide if a test requiring a minimum size should be run or not?

What if we had a field inside `TestConfig` to indicate the test's required size range, and also have a `GetClaimSizeRange` returned by the driver? `TestConfig` will then also get passed into `setupResources()`. Upon volume creation the framework always picks the minimum of the intersection between the test's range and the driver's range.

This way we don't artificially limit potential test <-> driver matches just because the default sizes don't work out

pohly replied:
I had tried to implement that, but quickly ran into problems because the code that uses the size doesn't always (or never? don't remember exactly) know which size the test that is going to run will need.

Can we leave solving that problem for the future? Taking the size from the driver definition is an improvement over hard-coding it in the tests, even if it isn't perfect.

pohly replied:
Done, see https://github.com/kubernetes/kubernetes/issues/72080

pohly commented:
@msau42 what do you think about @mkimuram's proposal with a separate `lib` directory?

Do you have a preference whether `TestDriver` should be defined in its own package (this PR) or inside `testsuites`? I don't have a strong opinion either way. I did it this way with the hope that a driver could be implemented by just pulling in the `testsuites/testdriver` package, but in practice the current drivers also still depend on some code from `testsuites` (specifically, `testsuites.GetUniqueDriverName` and `testsuites.GetStorageClass`). I'm now leaning towards simpy moving `testsuites/testdriver/testdriver.go` to `testsuites/testdriver.go`.

mkimuram commented:
@pohly 

Thank you for your comment.

> Is it supposed to be generic test execution logic that could also be used for some entirely different domain, say test/e2e/lifecycle? That can't be achieved just by moving around files, because for example lib/testsuites/base.go would still be specific to the test suites defined in testsuites.

Correct, it can only handle test suites. If this matters, you can move it, like moving entire lib package under storage/testsuites.

> What would be the name of the package? lib is much to generic.

You can rename it as you like.

My purposes were:
1. Keeping this framework usable from out-of-tree,
2. Keeping package layout easy to follow.

To achieve 1:
- We might be able to separate definition and implementation for testsuites.
- We might be able to prepare a common logic to be called in the same way from both k/k test and out-of-tree test.

To achieve 2:
- We might be able to move definition under the same package

The package layout I detailed was just an example to achieve above purposes by above way, and I didn't intend to force it.

msau42 commented:
I don't have a strong preference, maybe slightly leaning towards a single package for simplicity.  Is there any situation where you would want to import one and not the other?

pohly commented:
@mkimuram there's no need for a separate `lib` package to achieve goal 1 - this PR shows that it is already usable without that. So for the sake of goal 2 (keeping it simple), I prefer to not create a `lib` package.

I agree that `testsuites/testdriver` isn't needed and will merge it back into `testsuites`.

mkimuram commented:
@pohly 

> this PR shows that it is already usable without that.

IIUC, you just tried to define another implementation of ```TestDriver``` out-of-tree.
What happens if you define another  implementation of ```TestSuite``` out-of-tree? Does it still work?

Also, how will we change and add tests in k/k without affecting ```TestDriver``` and ```TestSuite``` used out-of-tree? (At least we should try to minimize the side effect of the change in upstream.) Does this PR achieve this?

pohly commented:
@mkimuram defining test drivers out-of-tree works. One example is public in `csi-e2e`, another one is something that I have internally.

How well it works as the API evolves depends on how careful we are with the API. My position is that we shouldn't break out-of-tree drivers by being careless, but we also shouldn't be too worried about it. As an out-of-tree user of the code I am expecting to update my test driver when importing a new revision of the code.

I have not tried defining a new testsuite of-of-tree. This wasn't what I had in mind, and I also don't know who would be interested in doing that. I expect problems with it because the test driver interface to some extend is written so that it has the necessary information for the current test suites - a new test suite might need settings that aren't supported.

mkimuram commented:
@pohly 

>  This wasn't what I had in mind, and I also don't know who would be interested in doing that.

If ```TestSuite``` can be defined out of tree, you will be able to define 1Gi claim test case just for your driver, without waiting for k/k change. It will be one of the use case.
 
To define ```TestSuite``` out-of-tree, I think that we should separate definition and implementation of ```TestSuite```. Then, similar approach for moving ```drivers``` package under ```testsuites``` package become inconsistent, like below:

```
storage/drivers/  ... implementation of drivers
storage/testsuites/ ... implementation of testsuites
storage/testsuites/drivers ... definition of drivers
storage/testsuites/testsuites? ... definition of testsuite
```

> I expect problems with it because the test driver interface to some extend is written so that it has the necessary information for the current test suites

Right. We should prepare document, once it become stable.

But just for a simple modification like above case,  just copy&paste them to out-of-tree and modify it should be fine. (There would be an issue related to access scope in reality, though. And to avoid this issue separation of definition and implementation of ```TestSuite``` should be considered.)

pohly commented:
@mkimuram, @msau42: I've rebased on top of current master and moved the `TestDriver` definition into the `testsuites` package.

I added an additional comment to `GetClaimSize`. I'd prefer to merge that as it is now, because it is an improvement over the current state (hard-coded value of 5Gi). If someone sees a need to make that more flexible and knows how, then that can be added in a future PR.

pohly commented:
Masaki Kimura <notifications@github.com> writes:
> If ```TestSuite``` can be defined out of tree, you will be able to
> define 1Gi claim test case just for your driver, without waiting for
> k/k change. It will be one of the use case.

Hmm, what's the point of having these shared tests if someone has to
write their own version whenever they want to change some parameter?

If someone wants to write a custom test, they can already so now. To me,
the value of the `testsuites` package is the tests, not the mechanism
for running them. By making them usable for out-of-tree drivers I hope
to encourage developers to contribute to the shared tests instead of
doing their own thing that doesn't benefit anyone else.

If someone wants to refactor further to make the writing of out-of-tree
test suites easier, then I won't object, but let's not hold back one
improvement for the sake of another that's further out and might not be
needed at all.

mkimuram commented:
> Hmm, what's the point of having these shared tests if someone has to
> write their own version whenever they want to change some parameter?

I meant just one example of use cases for adding or modifying ```TestSuites```.
And I just said that you can workaround it until it is fixed in k/k. 

There will be more use cases like: 
  - doing some extra tests that are specific to some driver
  - running some extra tests that are not yet merged in k/k until it is merged
  - adding more combination of tests that are not accepted in k/k
 
> If someone wants to refactor further to make the writing of out-of-tree
test suites easier, then I won't object, but let's not hold back one
improvement for the sake of another that's further out and might not be
needed at all.

As PR title is "e2e storage tests: usable out-of-tree", I meant to focus this topic and how to improve it to make it usable out-of-tree, in general.

However, IIUC, this PR is doing two things:
1. Making Claim size tunable for each driver
2. Reducing the dependency packages that is required

If above are separated PR and you are only talking about 1, "not holding back on improvement for the sake of another" makes sense. It seems a blocker for you to make it work.
(Just for 1, we can change back claim size to 1Gi, if needed, once latest csi gce test become used. )

However, without 2, it is already usable out-of-tree thanks to your work, although the size of dependency packages are big. So, merging both of them at once just for prioritizing 1 doesn't sound good to me.

For 2, I think that we should consider more about the use case and decide how we should do. Otherwise, we would need to move packages back and force many times.

pohly commented:
Masaki Kimura <notifications@github.com> writes:
> As PR title is "e2e storage tests: usable out-of-tree", I meant to
> focus this topic and how to improve it to make it usable out-of-tree,
> in general.

Exactly, this is about the "e2e storage *tests*" ;-} With that I meant
the existing tests, not the mechanism how they are written.

> However, IIUC, this PR is doing two things:
> 1. Making Claim size tunable for each driver
> 2. Reducing the dependency packages that is required
>
> If above are separated PR and you are only talking about 1, "not
> holding back on improvement for the sake of another" makes sense.

I was talking about both, see below.

> It seems a blocker for you to make it work.
> (Just for 1, we can change back claim size to 1Gi, if needed, once
> latest csi gce test become used. )

1Gi is too big for my test setup. I want 1Mi.

> However, without 2, it is already usable out-of-tree thanks to your
> work, although the size of dependency packages are big. So, merging
> both of them at once just for prioritizing 1 doesn't sound good to me.
>
> For 2, I think that we should consider more about the use case and
> decide how we should do. Otherwise, we would need to move packages
> back and force many times.

I consider the `testsuites` package unusable for out-of-tree CSI driver
tests suites if it ends up adding vsphere tests to those test
suites. This is of course debatable, but that's my opinion.

I don't see a problem with moving packages once more. No-one pulling
from k/k should expect API stability, and actual moving code is mostly a
matter of "git mv" and "sed", so it isn't much work either.

pohly commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce-100-performance

mkimuram commented:
@pohly 

O.K. I understand that you are only focusing on making your specific driver test work out-of-tree in your specific use case in this PR. I won't block this, if other people agree.

I will consider improving on top of this PR later.

pohly commented:
@msau42 do you agree to go ahead with this?

msau42 commented:
Sorry I will try to take a look at this once we're done with 1.13 release issues.  I agree with the direction.

pohly committed:
e2e/storage: remove code duplication

The setup of the V0 hostpath driver was done with copy-and-paste and
then changing just the driver name and the manifests. The same can be
achieved by making the base struct a bit more configurable, which
simplifies future changes (less code).

Renaming the provisioner container was unnecessary and was reverted to
make it possible to use the same patch configuration.

While at it, also fix the InitHostV0PathCSIDriver typo.

pohly commented:
/test pull-kubernetes-integration
/test pull-kubernetes-verify
/test pull-kubernetes-e2e-kops-aws

pohly commented:
/test pull-kubernetes-verify
/test pull-kubernetes-e2e-kops-aws

pohly commented:
/test pull-kubernetes-verify

msau42 commented:
/assign @verult

pohly commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -44,160 +44,111 @@ import (
-		return utils.PatchCSIDeployment(h.driverInfo.Framework, o, item)
-	},
+// InitHostPathV0CSIDriver returns a variant of hostpathCSIDriver with different manifests.
+func InitHostV0PathCSIDriver(config testsuites.TestConfig) testsuites.TestDriver {
```
I hadn't noticed that typo because I only modified the surrounding code. I agree, this should be fixed.

pohly replied:
Done, force-pushed.

pohly commented:
@verult I know, it's KubeCon, but life goes on ;-} Did I address your concerns? Can we get this merged now?

verult commented:
ACK, sorry for the delay, will take a look later today!

verult commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -1219,7 +1250,7 @@ func (g *gcePdDriver) CreateVolume(volType testpatterns.TestVolType) interface{}
 	if volType == testpatterns.InlineVolume {
 		// PD will be created in framework.TestContext.CloudConfig.Zone zone,
 		// so pods should be also scheduled there.
-		g.driverInfo.Config.NodeSelector = map[string]string{
```
Can some fields in `VolumeTestConfig` be removed (e.g. `NodeSelector`)?

pohly replied:
The GCE PD driver relies on `NodeSelector` and I was also using it out-of-tree, so no, I don't think it can be removed.

verult commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -99,9 +99,42 @@ type DriverInfo struct {
+
+// TestConfig represents parameters that control test execution.
+// They can still be modified after defining tests, for example
+// in a BeforeEach or when creating the driver.
```
My understanding is that `TestConfig` is ultimately used to configure each test, but it's passed to the driver during initialization so the driver has a chance to add more configs, is this correct?

It'd be helpful to define even more precisely how `TestConfig` is used here, to make its misuse by future devs less likely.

pohly replied:
Correct. `DriverInfo` is the static part that gets defined by the test driver author once. `TestConfig` is the part which is more dynamic and contains fields that can be different for different tests.

pohly replied:
@verult: Does 0fa0ac7ff75987c17a96c62bb4e00c6020cf8cd2 sufficiently explain the relationship between test definition, test execution, DriverInfo and TestConfig?

I am aware that `testsuites.RunTestSuites` is named a bit inappropriately, because it doesn't actually run any tests, it merely defines them. But I didn't want to make more changes than strictly necessary, therefore I have kept the name.

verult replied:
Thanks for the doc! Looks great to me. My one nit would be I don't think think it's necessary to describe the ginkgo interactions common across all e2es, but I don't mind either way.

pohly replied:
Let's keep it in. I'm going to come back to that in the next PR where I introduce delayed test registration, because then it won't be just the usual "register at init".

verult commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -99,9 +99,42 @@ type DriverInfo struct {
+// TestConfig represents parameters that control test execution.
+// They can still be modified after defining tests, for example
+// in a BeforeEach or when creating the driver.
+type TestConfig struct {
```
The existence of two `TestConfig`s is going to be very confusing. Is it possible to have a clearer divide between the purposes of the two structs?

Some possibilities:
* Improve the comments for `VolumeTestConfig` clearly describing its purpose, and possibly renaming it.
  * Most fields in `VolumeTestConfig` are only used in `volume_util.go`. Could we limit its use only in that file?
* Reduce (or ideally completely eliminate) field overlap between the two structs.

pohly replied:
As far as I understand it, `VolumeTestConfig` is from legacy code that @mkimuram continued to use unmodified because some tests and utility code rely on it. I agree that it would be useful to clean that up, but I barely know that code and would prefer to not touch it, or at least not now.

I'll try to enhance the documentation for the new struct, but when it comes to `VolumeTestConfig` I'll basically have to tell developers that they'll have to read the legacy code.

pohly replied:
See 0fa0ac7.

verult replied:
OK, I could create an issue to track the cleanup between the two configs in the future.

pohly replied:
https://github.com/kubernetes/kubernetes/issues/72242

verult reviewed (commented):
I'm not very sold on introducing a new `TestConfig` in its current form, but I agree with the spirit that `VolumeTestConfig` seems too broad. Everything else looks good.

pohly commented:
/test pull-kubernetes-integration

verult commented:
/lgtm

pohly commented:
I had to rebase because my other PR for the image versions changed the .yaml files.

@verult can you re-add your lgtm or
@msau42 can you lgtm+approve?

verult commented:
/lgtm

pohly commented:
@msau42 I took the enhancements for multiple nodes out of this PR and re-implemented that in https://github.com/kubernetes/kubernetes/pull/72002. That PR is now a bit bigger than it needs to be, let me know if you want me to split it up.

Please have another look at the remaining code in this PR.

pohly commented:
/test pull-kubernetes-integration

liggitt commented:
linking PR's that had a test flake because of https://github.com/kubernetes/kubernetes/issues/71696

msau42 commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -144,11 +146,14 @@ func (n *nfsDriver) GetClaimSize() string {
 	ns := f.Namespace
 	n.externalPluginName = fmt.Sprintf("example.com/nfs-%s", ns.Name)
 
+	// Reset config. It might have been modified by a previous CreateVolume call.
```
Tests can run in parallel. Does each test case get it's own unique object, or share the same object?  If each test case has a new object, then this is not needed.  If they share the same object, then this may not work very well.

msau42 replied:
Opened up https://github.com/kubernetes/kubernetes/issues/72288 to consider refactoring this later.  I don't think this is a clean abstraction

pohly commented:
Michelle Au <notifications@github.com> writes:
> +	// Reset config. It might have been modified by a previous CreateVolume call.
>
> Tests can run in parallel.

But only in different processes. This is a difference between Gingko and
"go test".

> Does each test case get it's own unique object, or share the same
> object?

Each process gets its own unique object, which is then shared between
all tests running in that process.

msau42 commented:
/lgtm
/approve

msau42 commented:
/priority important-soon
